File: typescript/src/stack/stack.ts
// vim: noexpandtab

export class EmptyStack extends Error {
	constructor(message?: string) {
		super(message);
	}
}

export class SNode<T> {
	value: T;
	prev?: SNode<T>;

	constructor(value: T) {
		this.value = value;
		this.prev = undefined;
	}
}

export class Stack<T> {
	public length: number;
	private head?: SNode<T>;

	constructor() {
		this.length = 0;
		this.head = undefined;
	}

	/**
	 *
	 */
	push(item: T) {
		const node = new SNode(item);

		this.length++;

		if (!this.head) {
			this.head = node;
			return;
		}

		node.prev = this.head;
		this.head = node;
	}

	/**
	 *
	 */
	pop(): T {
		if (!this.head) {
			throw new EmptyStack('Stack is empty');
		}

		this.length--;
		const head = this.head;
		this.head = head.prev;

		return head.value;
	}

	/**
	 * Returns value of stack without mutation
	 * @throws {EmptyStack} - If/when stack is empty
	 */
	peek(): T {
		if (this.head) {
			return this.head.value;
		}
		throw new EmptyStack('Stack is empty');
	}
}

File: go/stack/stack.go
package stack

import "errors"

type Node[T any] struct {
	value T
	prev  *Node[T]
}

type Stack[T any] struct {
	Length uint
	head   *Node[T]
}

/**
 *
 */
func (stack *Stack[T]) Push(item T) {
	node := Node[T]{
		value: item,
	}

	stack.Length++

	if stack.Length == 1 {
		stack.head = &node
		return
	}

	node.prev = stack.head
	stack.head = &node
}

/**
 *
 */
func (stack *Stack[T]) Pop() (T, error) {
	if stack.Length == 0 {
		var result T
		return result, errors.New("Stack is empty")
	}

	stack.Length--
	head := stack.head
	stack.head = head.prev

	return head.value, nil
}

/**
 * Returns first value of stack without mutation
 */
func (stack *Stack[T]) Peek() (T, error) {
	if stack.Length == 0 {
		var result T
		return result, errors.New("Queue is empty")
	}

	return stack.head.value, nil
}
